File: 2915-count-of-interesting-subarrays/2915-count-of-interesting-subarrays.java
class Solution {
    public long countInterestingSubarrays(List<Integer> nums, int modulo, int k) {
        int n = nums.size();
        Map<Integer,Long> mp = new HashMap<>();
        mp.put(0,1L);

        long sm = 0;
        long res = 0;

        for(int i = 0; i<n; i++){
            if(nums.get(i) % modulo == k){
                sm += 1;
            }
            int r1 = (int)sm % modulo;
            int r2 = (r1 - k + modulo) % modulo;

            res += mp.getOrDefault(r2, 0L);
            mp.put((int)r1, mp.getOrDefault(r1,0L)+1);
        }
        return res;
    }
}

File: 2915-count-of-interesting-subarrays/2915-count-of-interesting-subarrays.cpp
class Solution {
public:
    long long countInterestingSubarrays(vector<int>& nums, int modulo, int k) {
        int n = nums.size();

        long long res = 0;
        unordered_map<int,long long> mp; // store reminder => count
        mp[0] = 1;
        long long sum = 0;

        for(int i = 0; i<n; i++){ 
            if(nums[i] % modulo == k)  
                sum += 1; 
            int r1 = sum % modulo;
            int r2 = (r1-k + modulo)% modulo;
            res += mp[r2];
            mp[r1]++;
        }
        return res;
    }
};